File: server/csharp/MetaObjects.Render/Extract/ExtractSchema.cs
namespace MetaObjects.Render.Extract;

/// <summary>
/// Top-level extract descriptor.
/// <see cref="RootName"/> is the XML root element tag / logical JSON root name.
/// <see cref="Fields"/> is never null; a null argument is normalised to an empty list.
/// </summary>
public sealed record ExtractSchema
{
    public Format Format { get; init; }
    public string RootName { get; init; }
    public IReadOnlyList<FieldSpec> Fields { get; init; }

    public ExtractSchema(Format format, string rootName, IReadOnlyList<FieldSpec>? fields = null)
    {
        Format = format;
        RootName = rootName;
        Fields = fields ?? Array.Empty<FieldSpec>();
    }
}

File: server/java/render/src/main/java/com/metaobjects/render/extract/ExtractSchema.java
package com.metaobjects.render.extract;

import java.util.List;
import java.util.Objects;

/** Top-level extract descriptor. rootName = the XML root tag / logical JSON root name. */
public record ExtractSchema(Format format, String rootName, List<FieldSpec> fields) {
    public ExtractSchema {
        Objects.requireNonNull(format, "format");
        Objects.requireNonNull(rootName, "rootName");
        fields = fields == null ? List.of() : List.copyOf(fields);
    }
}
